language/stdlib/modules/DesignatedDealer.move

```diff
@@ -185,10 +229,18 @@ module DesignatedDealer {
     spec fun update_tier {
         pragma opaque = true;
         modifies global<TierInfo<CoinType>>(dd_addr);
+        ensures exists<TierInfo<CoinType>>(dd_addr);
         ensures global<TierInfo<CoinType>>(dd_addr).window_start == old(global<TierInfo<CoinType>>(dd_addr)).window_start;
         ensures global<TierInfo<CoinType>>(dd_addr).window_inflow == old(global<TierInfo<CoinType>>(dd_addr)).window_inflow;
         ensures len(global<TierInfo<CoinType>>(dd_addr).tiers) == len(old(global<TierInfo<CoinType>>(dd_addr)).tiers);
         ensures global<TierInfo<CoinType>>(dd_addr).tiers[tier_index] == new_upperbound;
+
+        include Roles::AbortsIfNotTreasuryCompliance{account: tc_account};
+        aborts_if !exists<TierInfo<CoinType>>(dd_addr);
+        let tier_info = global<TierInfo<CoinType>>(dd_addr);
+        aborts_if tier_index >= len(tier_info.tiers) with Errors::INVALID_ARGUMENT;
+        aborts_if tier_index > 0 && tier_info.tiers[tier_index - 1] >= new_upperbound with Errors::INVALID_ARGUMENT;
+        aborts_if tier_index + 1 < len(tier_info.tiers) && tier_info.tiers[tier_index + 1] <= new_upperbound with Errors::INVALID_ARGUMENT;
```

Comment:
Was just thinking about this, and wondering if there is any way of proving that only the tier at `tier_index` is changed, and all others remain the same?

Reply:
I have made the spec of update_tier and add_tier more precise.